src/internal/logs/matchlines.go

```diff
@@ -0,0 +1,206 @@
+package logs
+
+import (
+	"encoding/json"
+	"regexp"
+	"time"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errutil"
+)
+
+type LineMatcher func(line []byte, fields *LogFields) bool
+
+// MatchLines implements LineMatcher interface
+type MatchLines struct {
+	matchers [](func(line []byte, fields *LogFields) bool)
+	Matchers [](func(line []byte, fields *LogFields) bool)
+}
+
+type LogFields struct {
+	Ts           time.Time
+	Level        string
+	ProjectName  string
+	PipelineName string
+	JobId        string
+	WorkerId     string
+	DatumId      string
+	Master       bool
+	User         bool
+}
+
+func (m *MatchLines) Match(line string, returnMatch func([]byte) error) error {
+
+	logFields := new(LogFields)
+	json.Unmarshal([]byte(line), logFields)
+
+	for _, matcher := range m.Matchers {
+		if !matcher([]byte(line), logFields) {
+			continue
+		}
+		if err := returnMatch([]byte(line)); err != nil {
+			if errors.Is(err, errutil.ErrBreak) {
+				return nil
+			}
+			return err
+		}
+	}
+	return nil
+}
+
+func TrueMatcher() (func(line []byte, fields *LogFields) bool, error) {
+	return func(line []byte, fields *LogFields) bool {
+		return true
+	}, nil
+}
+
+func FalseMatcher() (func(line []byte, fields *LogFields) bool, error) {
+	return func(line []byte, fields *LogFields) bool {
+		return false
+	}, nil
+}
+
+func TimeMatcher(fromTime time.Time, untilTime time.Time) (func(line []byte, fields *LogFields) bool, error) {
+	if fromTime.After(untilTime) {
+		return nil, errors.Errorf("invalid parameters, start time %t is after end time %t", fromTime, untilTime)
```

Comment:
%t formats a boolean.  Use %s and fromTime.Format(time.RFC3339Nano) or similar.  (%v is the "go to" default formatter, but it's especially ugly for times.)